carbonmark-api/src/utils/helpers/fetchMarketplaceListings.ts

```diff
@@ -71,52 +73,74 @@ export const fetchMarketplaceListings = async ({
       : null,
   }));
 
-  const getListingsWithProfiles = Promise.all(
-    formattedListings.map(async (listing) => {
-      const seller = await fastify.firebase
-        .firestore()
-        .collection("users")
-        .doc(listing.seller.id.toUpperCase())
-        .get();
-      return {
-        ...listing,
-        seller: {
-          ...listing.seller,
-          ...seller.data(),
-        },
-      };
-    })
+  const userIds = new Set<string>();
+  formattedListings.forEach((listing) =>
+    userIds.add(listing.seller.id.toLowerCase())
   );
-  const getActivitiesWithProfiles: Promise<ActivityWithUserHandles[]> =
-    Promise.all(
-      formattedActivities.map(async (act) => {
-        const activityWithHandles: ActivityWithUserHandles = { ...act };
-        const seller = await fastify.firebase
+  formattedActivities.forEach((activity) => {
+    userIds.add(activity.seller.id.toLowerCase());
+    if (activity.buyer) {
+      userIds.add(activity.buyer.id.toLowerCase());
+    }
+  });
+
+  const usersById = new Map<string, DocumentData | undefined>();
+
+  if (userIds.size !== 0) {
+    const chunkSize = 30;
+    const chunks = [];
+    const ids = Array.from(userIds);
+
+    for (let i = 0; i < ids.length; i += chunkSize) {
+      chunks.push(ids.slice(i, i + chunkSize));
+    }
+
+    const callStart = Date.now();
+    // @todo this requires that users in production has an address field. Staging is updated to reflect this. Currently the id is the only location of the address.
+    const userDocs = await Promise.all(
+      chunks.map((chunk) =>
+        fastify.firebase
           .firestore()
           .collection("users")
-          .doc(act.seller.id.toUpperCase())
-          .get();
-        if (seller.exists) {
-          activityWithHandles.seller.handle = seller.data()?.handle;
-        }
-        if (act.buyer) {
-          const buyer = await fastify.firebase
-            .firestore()
-            .collection("users")
-            .doc(act.buyer.id.toUpperCase())
-            .get();
-          if (buyer.exists) {
-            assign(activityWithHandles, "buyer.handle", buyer.data()?.handle);
-          }
-        }
-        return activityWithHandles;
-      })
+          .where("address", "in", chunk)
+          .get()
+      )
     );
```

Comment:
This is doing a network call for each item in the chunk array, correct? If so, I think we should reduce it to a single call given that the `in` operator seems to support array values.

See [here](https://firebase.google.com/docs/firestore/query-data/queries?hl=de#in_and_array-contains-any) for an example. What do you think?

Reply:
I believe it's making one network call per `chunk` array; firestore has a 30 value limit per chunk (if you include more values in the array you get this error from firestore:  `details: "'IN' supports up to 30 comparison values."`).  So a 75 value `chunks` array would be split up into 3 `chunk` arrays: 30, 30, 15 for a minimum of three calls. If I'm incorrect though it would be great to reduce the number of calls(!)